tools/src/main/scala/scala/scalanative/compiler/pass/GlobalValueNumbering.scala

```diff
@@ -2,11 +2,283 @@ package scala.scalanative
+      cfg: ControlFlow.Graph,
+      domination: Map[Block, Set[Block]]): Seq[Inst] = {
+
+    val variableVN   = mutable.HashMap.empty[Local, Hash]
+    val instructions = mutable.HashMap.empty[Hash, List[Inst.Let]]
+    val localDefs    = mutable.HashMap.empty[Local, Inst]
+
+    val hash       = new HashFunction(variableVN)
+    val deepEquals = new DeepEquals(localDefs)
+
+    def blockDominatedByDef(dominatedBlock: Block,
+                            dominatingDef: Local): Boolean = {
+
+      domination(dominatedBlock).exists { dominatingBlock =>
+        val foundInParam = dominatingBlock.params.exists {
+          case Val.Local(paramName, _) => (paramName == dominatingDef)
+        }
+        val foundInInsts = dominatingBlock.insts.exists {
+          case Inst.Let(name, _) => (name == dominatingDef)
+          case _                 => false
+        }
+
+        foundInParam || foundInInsts
+      }
+    }
+
+    val newInsts = cfg.map { block =>
+      variableVN ++= block.params.map(lval =>
+        (lval.name, HashFunction.rawLocal(lval.name)))
+      localDefs ++= block.params.map(lval => (lval.name, block.label))
+
+      val newBlockInsts = block.insts.map {
+
+        case inst: Inst.Let => {
+          val idempotent = isIdempotent(inst.op)
+
+          val instHash =
+            if (idempotent)
+              hash(inst.op)
+            else
+              inst.hashCode // hash the assigned variable as well, so a = op(b) and c = op(b) don't have the same hash
+
+          variableVN += (inst.name -> instHash)
+          localDefs += (inst.name  -> inst)
+
+          if (idempotent) {
+            val hashEqualInstrs = instructions.getOrElse(instHash, Nil)
+            instructions += (instHash -> (inst :: hashEqualInstrs))
+
+            val equalInstrs =
+              hashEqualInstrs.filter(otherInst =>
+                deepEquals.eqInst(inst, otherInst))
+            val redundantInstrs = equalInstrs.filter(eqInst =>
+              blockDominatedByDef(block, eqInst.name)) // only redundant if the current block is dominated by the block in which the equal instruction occurs
+
+            val newInstOpt = redundantInstrs.headOption.map(
+              redInst =>
+                Inst.Let(inst.name,
+                         Op.Copy(Val.Local(redInst.name, redInst.op.resty))))
+            newInstOpt.getOrElse(inst)
+          } else {
+            inst
+          }
+        }
+
+        case otherInst @ _ =>
+          otherInst
+      }
+
+      block.label +: newBlockInsts
+    }
+
+    newInsts.flatten
+  }
+
+}
+
+object GlobalValueNumbering extends PassCompanion {
+  def apply(ctx: Ctx) = new GlobalValueNumbering()(ctx.top)
+
+  /* The definition of idempotent for Call might be wrong.
+   * Example:
+   * val a = new AnyRef { var b: Int = 0; override def toString = b.toString }
+   * val c = a.toString
+   * a.b = 1
+   * val d = a.toString
+   * --> toString is pure, but the values of c and d are different !
+   * */
```

Comment:
`toString` is not pure, `load`-s on external memory are side-effecting by definition.